Sources/SentryCrash/Recording/SentryCrashBinaryImageCache.c

```diff
@@ -0,0 +1,156 @@
+#include "SentryCrashBinaryImageCache.h"
+#include "SentryCrashDynamicLinker.h"
+#include <mach-o/dyld.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#if TEST || TESTCI
+
+typedef void (*SentryRegisterImageCallback)(const struct mach_header *mh, intptr_t vmaddr_slide);
+typedef void (*SentryRegisterFunction)(SentryRegisterImageCallback function);
+
+static SentryRegisterFunction _sentry_register_func_for_add_image
+    = &_dyld_register_func_for_add_image;
+static SentryRegisterFunction _sentry_register_func_for_remove_image
+    = &_dyld_register_func_for_remove_image;
+
+void
+sentry_setRegisterFuncForAddImage(SentryRegisterFunction addFunction)
+{
+    _sentry_register_func_for_add_image = addFunction;
+}
+
+void
+sentry_setRegisterFuncForRemoveImage(SentryRegisterFunction removeFunction)
+{
+    _sentry_register_func_for_remove_image = removeFunction;
+}
+
+void
+sentry_resetFuncForAddRemoveImage(void)
+{
+    _sentry_register_func_for_add_image = &_dyld_register_func_for_add_image;
+    _sentry_register_func_for_remove_image = &_dyld_register_func_for_remove_image;
+}
+
+#    define sentry_dyld_register_func_for_add_image(CALLBACK)                                      \
+        _sentry_register_func_for_add_image(CALLBACK);
+#    define sentry_dyld_register_func_for_remove_image(CALLBACK)                                   \
+        _sentry_register_func_for_remove_image(CALLBACK);
+
+#else
+#    define sentry_dyld_register_func_for_add_image(CALLBACK)                                      \
+        _dyld_register_func_for_add_image(CALLBACK)
+#    define sentry_dyld_register_func_for_remove_image(CALLBACK)                                   \
+        _dyld_register_func_for_remove_image(CALLBACK)
+#endif
+
+typedef struct SentryCrashBinaryImageNode {
+    SentryCrashBinaryImage image;
+    bool available;
+    struct SentryCrashBinaryImageNode *next;
+} SentryCrashBinaryImageNode;
+
+static SentryCrashBinaryImageNode rootNode = { 0 };
+static SentryCrashBinaryImageNode *tailNode = NULL;
+static pthread_mutex_t binaryImagesMutex = PTHREAD_MUTEX_INITIALIZER;
+
+static void
+binaryImageAdded(const struct mach_header *header, intptr_t slide)
+{
+    if (tailNode == NULL) {
+        return;
+    }
+
+    Dl_info info;
+    if (!dladdr(header, &info) || info.dli_fname == NULL) {
+        return;
+    }
+
+    SentryCrashBinaryImage binaryImage = { 0 };
+    if (!sentrycrashdl_getBinaryImageForHeader(
+            (const void *)header, info.dli_fname, &binaryImage)) {
+        return;
+    }
+
+    SentryCrashBinaryImageNode *newNode = malloc(sizeof(SentryCrashBinaryImageNode));
+    newNode->available = true;
+    newNode->image = binaryImage;
+    newNode->next = NULL;
+
+    pthread_mutex_lock(&binaryImagesMutex);
+    tailNode->next = newNode;
```

Comment:
`h`: I think we must check if the `tailNode` became `NULL` again here, as another thread could call `sentrycrashbic_stopCache` since the check above.
```suggestion
    if (tailNode == NULL) {
        return;
    }
    tailNode->next = newNode;
```

Reply:
Sound right